corehq/apps/app_manager/suite_xml.py

```diff
@@ -1645,24 +1676,60 @@ def get_datum_session_var(action, action_index):
 
                     target_module = get_target_module(action.case_type, last_action.details_module, True)
 
-                    e.datums.append(SessionDatum(
-                        id='product_id',
-                        nodeset=nodeset,
-                        value="./@id",
-                        detail_select=self.get_detail_id_safe(target_module, 'product_short')
-                    ))
+                    datums.append({
+                        'datum': SessionDatum(
+                            id='product_id',
+                            nodeset=nodeset,
+                            value="./@id",
+                            detail_select=self.get_detail_id_safe(target_module, 'product_short')
+                        ),
+                        'case_type': None,
+                        'requires_selection': True,
+                        'action': None
+                    })
             except IndexError:
                 pass
 
-        datums = self.get_new_case_id_datums(form)
-        for datum in datums:
-            e.datums.append(datum)
-
-        if form.is_registration_form() and form.is_case_list_form:
-            self.configure_entry_as_case_list_form(form, e)
-
-        if self.app.case_sharing and case_sharing_requires_assertion(form):
-            self.add_case_sharing_assertion(e)
+        root_module = module.root_module
+        root_datums = []
+        if root_module and root_module.module_type == 'basic':
+            try:
+                # assume that all forms in the root module have the same case management
+                root_module_form = root_module.get_form(0)
+            except FormNotFoundException:
+                pass
+            else:
+                if root_module_form.requires_case():
+                    root_datums.extend(self.get_datum_meta_module(root_module))
+                root_datums.extend(self.get_new_case_id_datums_meta(root_module_form))
+
+        if root_datums:
+            # we need to try and match the datums to the root module so that
+            # the navigation on the phone works correctly
+            # 1. Add in any datums that don't require user selection e.g. new case IDs
+            # 2. Match the datum ID for datums that appear in the same position and
+            #    will be loading the same case type
+            # see advanced_app_features#child-modules in docs
+            datum_pairs = izip_longest(datums, root_datums)
+            index = 0
+            for this_datum_meta, parent_datum_meta in datum_pairs:
+                if not this_datum_meta:
```

Comment:
minor but this could be combined with if statement below to: `if not this_datum_meta or not parent_datum_meta:` or `if not (this_datum_meta and parent_datum_meta):`